clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -4605,6 +4608,47 @@ void handleFailure(Throwable throwable) {
         return new UpdateFeaturesResult(new HashMap<>(updateFutures));
     }
 
+    @Override
+    public DecommissionBrokerResult decommissionBroker(int brokerId, DecommissionBrokerOptions options) {
+        final KafkaFutureImpl<Void> future = new KafkaFutureImpl<>();
+        final long now = time.milliseconds();
+        final Call call = new Call("decommissionBroker", calcDeadlineMs(now, options.timeoutMs()),
+            new LeastLoadedNodeProvider()) {
+
+            @Override
+            DecommissionBrokerRequest.Builder createRequest(int timeoutMs) {
+                DecommissionBrokerRequestData data =
+                    new DecommissionBrokerRequestData().setBrokerId(brokerId);
+                return new DecommissionBrokerRequest.Builder(data);
+            }
+
+            @Override
+            void handleResponse(AbstractResponse abstractResponse) {
+                final DecommissionBrokerResponse response =
+                    (DecommissionBrokerResponse) abstractResponse;
+                Errors error = Errors.forCode(response.data().errorCode());
+                switch (error) {
+                    case NONE:
+                        future.complete(null);
+                        break;
+                    case NOT_CONTROLLER:
+                        handleNotControllerError(error);
```

Comment:
Do we actually expect to get back `NOT_CONTROLLER`? I would have thought that this error would be swallowed by the broker when it forwards the request. In the current `ForwardingManager`, the broker will try to send the request to the controller. If the controller returns `NOT_CONTROLLER`, it will just retry. It will continue doing so until it ultimately times out. So maybe `REQUEST_TIMED_OUT` is the one we should be looking for?

Reply:
Hmm. Good point. We don't need a special-case for the REQUEST_TIMED_OUT then. Will update the test and case here